src/redisearch.h

```diff
@@ -42,19 +43,28 @@ typedef enum {
   Document_HasOffsetVector = 0x08
 } RSDocumentFlags;
 
+typedef struct {
+  const char *keyStr;
+  size_t keyLen;
+} RSDocumentKey;
+
+static inline RSDocumentKey MakeDocKey(const char *key, size_t len) {
+  return (RSDocumentKey){.keyStr = key, .keyLen = len};
+}
+
 /* RSDocumentMetadata describes metadata stored about a document in the index (not the document
-* itself).
-*
-* The key is the actual user defined key of the document, not the incremental id. It is used to
-* convert incremental internal ids to external string keys.
-*
-* Score is the original user score as inserted to the index
-*
-* Flags is not currently used, but should be used in the future to mark documents as deleted, etc.
-*/
+ * itself).
+ *
+ * The key is the actual user defined key of the document, not the incremental id. It is used to
+ * convert incremental internal ids to external string keys.
+ *
+ * Score is the original user score as inserted to the index
+ *
+ * Flags is not currently used, but should be used in the future to mark documents as deleted, etc.
+ */
 typedef struct {
   /* The actual key of the document, not the internal incremental id */
-  char *key;
+  sds keyPtr;
```

Comment:
We should hide this in an opaque data type, and wrap all access to it with functions, just like RediModuleString. Not because of cleanliness - but to avoid linking problems and symbol mix ups in redisearch external extensios.

Reply:
We're already using sds in query.{c,h}, so I don't see any harm done here. I do agree generally with the idea that we need to abstract this properly though, but adding more opacity isn't going to help us.

There is indeed a risk of symbol collision with Redis' own symbols though. Probably the best way would be to run our own copy of SDS through some text processing so it becomes rmutil_sds RSsds or similar.

Reply:
redisearch.h is **exposed to redisearch extensions**. Once you do that you're forcing them to use sds insteaf of calling our own symbols that wrap it.

Reply:
If the problem is simply hiding sds.h from users, I'll mark `keyPtr` as `char*` and do the sds stuff inside `doc_table.h` which I would assume is not intended for user exclusion.